Домашнее задание по 10 спринту

WayOfTheDev opened the pull request:
Добавление SplashViewController
Переход с экрана авторизации на экран с лентой

WayOfTheDev committed:
Домашнее задание по 10 спринту

Добавление SplashViewController
Переход с экрана авторизации на экран с лентой

andrew-kuzn-yndx commented on Image Feed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,116 @@
+final class AuthViewController: UIViewController, WebViewViewControllerDelegate {
+    
+    // MARK: - Outlets
+    @IBOutlet weak var loginButton: UIButton!
```
Молодец, что указал weak! Также не забывай также указание модификатора private к IBOutlet

andrew-kuzn-yndx commented on Image Feed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,116 @@
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        print("AuthViewController: prepare(for:sender:) вызван")
+        print("Тип segue.destination: \(type(of: segue.destination))")
+        print("Полное имя класса segue.destination: \(NSStringFromClass(type(of: segue.destination)))")
```
Старайся не оставлять такое, правилом хорошего тона является чистить код от лишнего перед сдачей работы. В основную ветку проекта должен попадать только чистый код без лишних принтов и закомментированных блоков)

andrew-kuzn-yndx commented on Image Feed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,172 @@
+//        WKWebsiteDataStore.default().removeData(ofTypes: websiteDataTypes, modifiedSince: date) {
+//            print("Cookies and cache cleared")
+//            self.webView.load(request)
+//        }
```
Давай почистим весь проект от лишних принтов и закомментированных блоков

andrew-kuzn-yndx commented on Image Feed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,172 @@
+        if
+            let url = navigationAction.request.url,
+            let urlComponents = URLComponents(string: url.absoluteString),
+            urlComponents.path == "/oauth/authorize/native",
```
Давай строки path тоже поместим в файл Constants, который у тебя есть) используй его по полной

andrew-kuzn-yndx commented on Image Feed/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,81 @@
+    
+    // MARK: - Private Methods
+    private func makeTokenRequest(with code: String) -> URLRequest? {
+        guard var urlComponents = URLComponents(string: "https://unsplash.com/oauth/token") else {
```
Этот юрл тоже можно поместить в Constants

andrew-kuzn-yndx commented on Image Feed/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,81 @@
+            switch result {
+            case .success(let data):
+                do {
+                    let response = try JSONDecoder().decode(OAuthTokenResponseBody.self, from: data)
```
Для того, чтобы не работать с snake_case ключами, не описывать их в CodingKeys, ты можешь также настраивать JSONDecoder под тот вид данных, который приходит от сервера. 
Рекомендую изучить этот способ самостоятельно https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable
В данном случае можно добавить настройку для декодера и избавиться в структуре OAuthTokenResponseBody от enum CodingKeys
Примерно такое решение:
```
                    let decoder = JSONDecoder()
                    decoder.keyDecodingStrategy = .convertFromSnakeCase
                    let response = try decoder.decode(OAuthTokenResponseBody.self, from: data)
                    OAuth2TokenStorage.shared.token = token
```

WayOfTheDev committed:
Fix bugs

1. Удалил ненужные print и закомментированный код
2. Перенес path и url в Constants
3. Удалил CodingKeys и учел комментарий по преобразованию snake_case ключи из JSON в camelCase.

Критичные правки по проекту:
1. Добавил Singleton в OAuth2Service, подправил зависимые блоки кода в AuthViewController и SplashViewController
2. Устранил проблему с сохранением данных авторизации, если быть точнее, то закомментировал строчку - oauth2TokenStorage.token = nil, так как с помощью нее делал проверку для очистки токена и повторной авторизации.
